grpclog/component.go

```diff
@@ -0,0 +1,289 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpclog
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+
+	"google.golang.org/grpc/internal/grpclog"
+)
+
+// Component is a unique identifier for a component. Created with NewComponent.
+type Component string
+
+// componentData records the settings for a component.
+type componentData struct {
+	verbosity int
+	level     int
+}
+
+const envName = "GRPC_GO_LOG_LEVEL"
+const sentinel = math.MinInt32
+
+// grpc components
+const (
+	CORE       = Component("CORE")
+	ALTS       = Component("ALTS")
+	GRPCLB     = Component("BALANCER_GRPCLB")
+	RLS        = Component("BALANCER_RLS")
+	ROUNDROBIN = Component("BALANCER_ROUNDROBIN")
+	BINARYLOG  = Component("BINARYLOG")
+	CHANNELZ   = Component("CHANNELZ")
+	DNS        = Component("RESOLVER_DNS")
+	TRANSPORT  = Component("TRANSPORT")
+	XDS        = Component("XDS")
+)
+
+var components = map[Component]*componentData{}
+var grpcComponents = []Component{
+	CORE,
+	ALTS,
+	GRPCLB,
+	RLS,
+	ROUNDROBIN,
+	BINARYLOG,
+	CHANNELZ,
+	DNS,
+	TRANSPORT,
+	XDS,
+}
+var environmentVars = map[string]*componentData{}
+var prefixVars = map[string]*componentData{}
+
+// init creates all grpc components and any components named in the environment variable and applies all settings
+// specified in the environment variable. Notably, settings for all components with a given prefix are applies before
+// regular component settings.
+func init() {
+	// Initialize the grpc components
+	for _, c := range grpcComponents {
+		components[c] = &componentData{0, 0}
+	}
+	// Pull environment variable data
+	if envVar, ok := os.LookupEnv(envName); ok && len(envVar) > 0 {
+		varList := strings.Split(envVar, ",")
```

Comment:
Move string parsing into a function, and add tests.

Reply:
Moved all environment variable string parsing to a function (parseEnvironmentVar) and added test cases for the function in a testing file.